src/protocol/command_matcher.rs

```diff
@@ -0,0 +1,115 @@
+use super::command::Command;
+use super::executable_path;
+use crate::R;
+use regex::Regex;
+use std::str;
+
+#[derive(Debug, Clone)]
+pub struct AnchoredRegex(String, Regex);
+
+impl PartialEq for AnchoredRegex {
+    fn eq(&self, other: &AnchoredRegex) -> bool {
+        match (self, other) {
+            (AnchoredRegex(a, _), AnchoredRegex(b, _)) => a == b,
+        }
+    }
+}
+
+impl Eq for AnchoredRegex {}
+
+impl AnchoredRegex {
+    pub fn new(raw_regex: &str) -> R<AnchoredRegex> {
+        Ok(AnchoredRegex(
+            raw_regex.to_string(),
+            Regex::new(&format!("^{}$", raw_regex))?,
+        ))
+    }
+
+    pub fn is_match(&self, other: &str) -> bool {
+        self.1.is_match(other)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum CommandMatcher {
+    ExactMatch(Command),
+    RegexMatch(AnchoredRegex),
+}
+
+impl CommandMatcher {
+    pub fn matches(&self, received: &Command) -> bool {
+        match self {
+            CommandMatcher::ExactMatch(command) => {
+                executable_path::compare_executables(&command.executable, &received.executable)
+                    && command.arguments == received.arguments
+            }
+            CommandMatcher::RegexMatch(regex) => regex.is_match(&received.format()),
+        }
+    }
+
+    pub fn format(&self) -> String {
+        match self {
+            CommandMatcher::ExactMatch(command) => command.format(),
+            CommandMatcher::RegexMatch(AnchoredRegex(string, _)) => string.clone(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod command_matcher {
```

Comment:
In all these tests use the simplest examples that test what the description says it tests.